### PR TITLE
Tests for `CatTargetDriftAnalyzer`, `chi_stat_test` and code simplifications

### LINKED ISSUE #96: χ2 test usage in border cases
Beware that there are some border cases when evidently computes a statistic on categorial data columns:

Now consider these two cases:

```
def test_current_data_contains_one_class_less(self):
        reference = DataFrame({
            'column_name': ['a'] * 5 + ['b'] * 5 + ['c'] * 5
        })
        current = DataFrame({
            'column_name': ['a'] * 8 + ['b'] * 3
        })
        result = chisquare_stattest.chi_stat_test(reference['column_name'], current['column_name'])

def test_reference_data_contains_one_class_less(self):
    reference = DataFrame({
        'column_name': ['a'] * 5 + ['b'] * 5
    })
    current = DataFrame({
        'column_name': ['a'] * 8 + ['b'] * 3 + ['c'] * 5
    })
    result = chisquare_stattest.chi_stat_test(reference['column_name'], current['column_name'])
```

In the fist case the `p_value` will be 0 because (when checking the code):
```
f_exp = [5, 5, 5]
f_obs = [8, 3, 0]
chisquare(f_exp, f_obs)
Out: Power_divergenceResult(statistic=inf, pvalue=0.0)
```
In the second case the statistic is actually fine
```
f_exp = [5, 5, 0]
f_obs = [8, 3, 5]
chisquare(f_exp, f_obs)
Power_divergenceResult(statistic=7.458333333333333, pvalue=0.024012838199733478)
```

That confuses me somehow. I find it alarming when `current_data` does have more classes than the `reference_data`. It means, is see new labels in new data that I haven't considered in reference data!

Any thoughts?

### PR DESCRIPTION
**Pull request by burkovae**
- [x] Add tests for `CatTargetDriftAnalyzer`
- [x] Refactor
- [x] Ask a bunch of questions :-)
  - [x] What about `target_type` ?
  - [x] Tests with `options`

closes #96

### CODE REVIEW

**Comment by burkovae** [comment]
I suggest making `option_providers.get(...)` as a parameter of the `calculate` method itself. It makes the method better testable, readable, and more robust against changes. Please think about doing it for other analyzers as well.

(as mentioned in chat, deeply hidden `option_provides` that depend on reading `config.json` does make writing tests painful. To encourage testing, writing test should require minimal effort ✔)

**Comment by Liraim** [comment]
> I suggest making `option_providers.get(...)` as a parameter of the `calculate` method itself. It makes the method better testable, readable, and more robust against changes. Please think about doing it for other analyzers as well.
> 

This is hard topic for discussion, because I personally don't like both options, but I will try to describe why I did so:

My first attempt for `options_provider` was like you describe, but at the moment it made signatures across all code (analyzers, tabs, widgets, etc) much heavier than I want it to be. So I put `options_provider` as public field in base classes. 

This solution make readability a bit harder (as you do not know that you need to mock this until you use it). But it doesn't affect testability as `options_provider` can be set after creating required object. For example:
```python
analyzer = DataDriftAnalyzer('')
options_provider = OptionsProvider()
options_provider.add(DataDriftOptions(confidence=0.5))
analyzer.options_provider = options_provider
analyzer.(reference_data, current_data, column_mapping)
```

Doesn't have much difference in testability (except readability) than:
```python
analyzer = DataDriftAnalyzer('')
options_provider = OptionsProvider()
options_provider.add(DataDriftOptions(confidence=0.5))
analyzer.(reference_data, current_data, column_mapping, options_provider)
```

_**But**_ because we have some other thoughts about changing of `calculate` method, so probably I will consider to move options_provider to arguments once more.

> (as mentioned in chat, deeply hidden `option_provides` that depend on reading `config.json` does make writing tests painful. To encourage testing, writing test should require minimal effort ✔)

This part is a bit confusing. `options_provider` doesn't depends on `config.json`. `config.json` is a configuration file for Evidently CLI (the way to run our library as console command). For using `Options` in scripts or notebooks you should manually create required `Options` (eg `DataDriftOptions`).

**Review by Liraim — CHANGES_REQUESTED** [review]

**Comment by burkovae** [comment]
Awsome @Liraim , thanks for your review.

Let me just respond to you comments:

> This solution make readability a bit harder (as you do not know that you need to mock this until you use it). But it doesn't affect testability as options_provider can be set after creating required object. For example:

     analyzer = DataDriftAnalyzer('')
     options_provider = OptionsProvider()
     options_provider.add(DataDriftOptions(confidence=0.5))
     analyzer.options_provider = options_provider
     analyzer.(reference_data, current_data, column_mapping)

> Doesn't have much difference in testability (except readability) than:

     analyzer = DataDriftAnalyzer('')
     options_provider = OptionsProvider()
     options_provider.add(DataDriftOptions(confidence=0.5))
     analyzer.(reference_data, current_data, column_mapping, options_provider)

Both solutions are problematic **because** they require knowledge about the implementation details. Perhaps moving `options_provider` as a constructor argument of analyzer will at least make it obvious for a reader/contributer/tester like so

    class Analyzer:
		def __init__():
			self.options_provider = OptionsProvider()
				
	    ...

	class CatTargetDriftAnalyzer(Analyzer):
		def __init__(options=None):
		    super().__init__()
			if options is None:
				options = DataDriftOptions(confidence=0.5)
			self.options_provider.add(options)
					
	    ...
		
But also something like this may help as well:

	class CatTargetDriftAnalyzer(Analyzer):
		def __init__(options=None):
			if options is None:
				options = DataDriftOptions(confidence=0.5)
			options_provider = OptionsProvider.get_instance().add(options)
					
	    ...

In such way, a user of this class will know that options exist in the context. I think a solution can be drafted in another PR. Thanks four sharing your thought.

File: evidently/analyzers/cat_target_drift_analyzer.py [hunk 1]
```diff
@@ -1,69 +1,96 @@
...
+    """Categorical target drift analyzer.
+
+    Analyze categorical `target` and `prediction` distributions and provide calculations to the following questions:
+    Does the model target behave similarly to the past period? Do my model predictions still look the same?
+
+    For reference see https://evidentlyai.com/blog/evidently-014-target-and-prediction-drift
+    """
+
+    def calculate(self, reference_data: pd.DataFrame, current_data: pd.DataFrame,
+                  column_mapping: ColumnMapping, options: DataDriftOptions = None) -> dict:
```
[the remark this answers is not in the document]
Comment (burkovae) [reply evidently/analyzers/cat_target_drift_analyzer.py]: Mh... well it didn't break any code 🤔 (and smoke tests).

That would be an excellent test case. Can you add one in another PR?

File: evidently/analyzers/cat_target_drift_analyzer.py [hunk 2]
```diff
@@ -1,69 +1,96 @@
...
 from evidently.analyzers.base_analyzer import Analyzer
-from evidently.options import DataDriftOptions
 from evidently.analyzers.stattests import z_stat_test, chi_stat_test
-from evidently.analyzers.utils import process_columns
+from .utils import process_columns
+from .. import ColumnMapping
+from ..options import DataDriftOptions
+
+
+# TODO: document somewhere, that all analyzers are mutators, i.e. they will change
```
[the remark this answers is not in the document]
Comment (burkovae) [reply evidently/analyzers/cat_target_drift_analyzer.py]: I removed TODO keywords and moved the comment to the culprit. When you takle it in another PR at some point, we can safely remove the comment then. ✔

File: evidently/analyzers/cat_target_drift_analyzer.py [hunk 3]
```diff
@@ -1,69 +1,96 @@
...
+            A dictionary that contains some meta information as well as `metrics` for either target or prediction
+            columns or both. The `*_drift` column in `metrics` contains a computed p_value from tests.
+        """
+        options = options or DataDriftOptions()
         columns = process_columns(reference_data, column_mapping)
         result = columns.as_dict()
         target_column = columns.utility_columns.target
         prediction_column = columns.utility_columns.prediction
 
+        # TODO: consider replacing only values in target and prediction column, see comment above
```
[the remark this answers is not in the document]
Comment (burkovae) [reply evidently/analyzers/cat_target_drift_analyzer.py]: dito ✔

File: evidently/analyzers/cat_target_drift_analyzer.py [hunk 4]
```diff
@@ -1,69 +1,96 @@
...
 import numpy as np
+import pandas as pd
 
 from evidently.analyzers.base_analyzer import Analyzer
-from evidently.options import DataDriftOptions
 from evidently.analyzers.stattests import z_stat_test, chi_stat_test
-from evidently.analyzers.utils import process_columns
+from .utils import process_columns
+from .. import ColumnMapping
+from ..options import DataDriftOptions
```
[the remark this answers is not in the document]
Comment (burkovae) [reply evidently/analyzers/cat_target_drift_analyzer.py]: Done ✔

File: evidently/analyzers/cat_target_drift_analyzer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (burkovae) [reply evidently/analyzers/cat_target_drift_analyzer.py]: btw: i used your suggestions above in the tests. As you said `options` are something beyond the scope of this PR.

[the remark this answers is not in the document]
Comment (Liraim) [reply evidently/analyzers/cat_target_drift_analyzer.py]: Yes, this won't break code execution, but will break options passing through hierarchy of components. Unfortunately, there are no test for that yet (I'm planning to add it as soon as can).

**Review by Liraim — APPROVED** [review]
Looks good now.

**Comment by Liraim** [comment]
> I have a quetion: It seems to me that you never use `prediction_type` or `target_type`:
> image

As far as I know, this fields aren't used in Widgets directly, but they are returned as fields in `Sections` in JSON-format (so it can be used outside of our code base). So I won't recommend to delete it until required by some strong reasons.

**Comment by burkovae** [comment]
@Liraim , with reference to [a SO question](https://stackoverflow.com/a/15700228/1256837) I believe you can get rid of code that you actually don't use yourself.
1. right now evidently is under active development - which is awesome - and I think you can allow yourself to "break" things to be more efficient with the code base.
2. If things break, there is always an issue tracker :D ... and you actually get to know the users and their use-cases - in particular, why they need such a feature (which you can pack in a test case again 🤩)

Whatever decision you reach, I think in the long run, when you [focus on the core features](https://ben.balter.com/2016/07/21/removing-a-feature-is-a-feature/) you'll deliver more value faster. But that's just my 2 cents. 

Thank you!

**Comment by burkovae** [comment]
I think that's it for this PR. Happy to help out and thanks for the discussion and feeback!

**Comment by Liraim** [comment]
> @Liraim , with reference to [a SO question](https://stackoverflow.com/a/15700228/1256837) I believe you can get rid of code that you actually don't use yourself.
> 
> 1. right now evidently is under active development - which is awesome - and I think you can allow yourself to "break" things to be more efficient with the code base.
> 2. If things break, there is always an issue tracker :D ... and you actually get to know the users and their use-cases - in particular, why they need such a feature (which you can pack in a test case again 🤩)
> 

I'm agree with this arguments, and we trying to delete all dead code.
And about subject - I will investigate usage of this field, and probably will remove, but I don't want to stale PR because it can take some time to check.

**Pull request merged**